src/main/scala/viper/silver/ast/utility/Chopper.scala

```diff
@@ -0,0 +1,903 @@
+    *                If none, then all members that induce a proof obligation are verified.
+    * @param bound   Specifies upper bound on the number of returned programs.
+    *                If none, then maximum number of programs is returned.
+    * @param penalty Specifies penalty of merging programs.
+    * @return Chopped programs.
+    */
+  def chop(choppee: ast.Program)(
+    isolate: Option[ast.Member => Boolean] = None,
+    bound: Option[Int] = Some(1),
+    penalty: Penalty[Vertex] = Penalty.Default
+  ): Vector[ast.Program] = {
+    chopWithMetrics(choppee)(isolate, bound, penalty)._1
+  }
+
+  /**
+    * chops `choppee` into multiple Viper programs.
+    *
+    * @param choppee Targeted program.
+    * @param isolate Specifies which members of the program should be verified.
+    *                If none, then all members that induce a proof obligation are verified.
+    * @param bound   Specifies upper bound on the number of returned programs.
+    *                If none, then maximum number of programs is returned.
+    * @param penalty Specifies penalty of merging programs.
+    * @return Chopped programs and metrics.
+    */
+  def chopWithMetrics(choppee: ast.Program)(
+    isolate: Option[ast.Member => Boolean] = None,
+    bound: Option[Int] = Some(1),
+    penalty: Penalty[Vertex] = Penalty.Default
+  ): (Vector[ast.Program], Metrics) = {
+
+    val graph = ViperGraph.toGraph(choppee, isolate)
+    val (programs, metrics) = Cut.boundedCut(graph)(bound, penalty)
+    (programs flatMap (list => graph.unapply(list)), metrics)
+  }
+
+  case class Metrics(
+                      maxNumberOfParts: Int,
+                      timeSCC: Option[BigDecimal],
+                      timeCutting: BigDecimal,
+                      timeMerging: BigDecimal,
+                    )
+
+  private object Cut {
+
+    /**
+      * Returns a set of chopped programs where the number of programs is bounded by `bound`.
+      *
+      * @param graph   Graph representation of the Viper program.
+      * @param bound   Specifies upper bound on the number of returned programs.
+      *                If none, then maximum number of programs is returned.
+      * @param penalty Specifies penalty of merging programs.
+      * @return Set of programs.
+      */
+    def boundedCut[T](graph: ViperGraph)(
+      bound: Option[Int],
+      penalty: Penalty[Vertex]
+    ): (Vector[Set[Int]], Metrics) = {
+      require(bound.forall(_ > 0), s"Got $bound as the size of the cut, but expected positive number")
+
+      /**
+        * During intermediate steps, programs are represented as sorted lists of ints.
+        * This representation makes merging programs linear time, which is done by [[penaltyAndMerge]].
+        * Furthermore, by pairing a node with its contribution to the merge penalty,
+        * the merge penalty can, together with the merging, be computed in linear time, too.
+        */
```

Comment:
A nitpick here: maybe, it makes sense to introduce a type alias for `Set[Int]` and use that definition throughout this file instead. Furthermore, I would move this comment to be next to that type alias, instead of beind "hidden" inside a method